lib/gollum-lib/markups.rb

```diff
@@ -6,7 +6,7 @@ class Markup
         Kramdown::Document.new(content, :auto_ids => false, :input => "markdown").to_html
     }
 
-    register(:markdown,  "Markdown", :regexp => /md|mkdn?|mdown|markdown/)
+    register(:markdown,  "Markdown", :regexp => /md|mkdn?|mdown|markdown/, :h1 => /^(#\s+)(\w+.*)/)
```

Comment:
> I think something like Page#edited_page doesn't really fit gollum, since it doesn't play well with git -- it assumes we can edit the Page's content, then save it to the repo, but there really isn't anything like a persisting, unsaved-but-modified page due to git's architecture (different content = different blob). That makes for less than ideally OO code, but that's the price of having git as a backend. :)

I'm sorry but I don't understand what you tried to say with this comment :p.
It seems I confused you again, sorry... should've made up a complete example :(. Let me clarify:

I think it would be quite troublesome to implement the "edit section" feature for irb processes. As such, I thought this feature would be frontend-exclusive (correct me if I'm wrong). If that's the case, suppose a class (frontend) like this:

```ruby
module Gollum
    class EditSectionPageHandle
        def initialize(path)
            # version is supposed to be the latest because... what's the point of editing a section of a historic version of a page?
            @page_source = ...      # use gollum-lib to read the page
            ...
        end
    end
    
    def get_section(section_number)
        # as defined above, except that this bit:
        #
        # ... render etc.
        # real_sections = nokogiri_doc.get_all_sections   # return all H1s by a css selector, for instance
        # 
        # should be defined in gollum-lib and called from the frontend (a.k.a. here)
    end
    
    def save_edited_page(new_section_content)
        wiki = ...
        edited_page = edited_page(new_section_content)     # as defined above
        wiki.update_page(page, name, format, editeddata, commit = {})
    end
    
    private
    
    def tokenize
        # as defined above
    end
    
    def edited_page(new_section)
        # as defined above
    end
```

**My original point was: define a frontend class to provide the "edit section" feature :).**

This frontend class clearly uses backend facilities to provide the "edit section" feature to the frontend. There are no "git issues", except for locking the page against other edits, which we have to solve anyway because otherwise, either one of these hinders Gollum's efforts:
* other edits cause the line numbers, offsets or even the `before, section, after` strings not to be reliable,
* saving the edited section causes the other edits to be ignored.

BTW, as far as I know, Gollum fails to protect users against this already with its editing features. A simple solution would be to make these features accessible only for Gollum instances bound to `localhost` (personal Wikis)...

Now, if instances of this class were stored in Sinatra's (or Rack's, not sure atm) user sessions, the process of updating the page (from frontend's point of view) is simple (pseudocode):

```ruby
map '/save-edited-page/...' do
    handler = Sinatra.get_session_for_request.get_page_handle(page_path)
    handler.save_edited_page(...)
end
```